cpp/oneapi/dal/backend/primitives/distance/correlation_distance_misc_dpc.cpp

```diff
@@ -0,0 +1,95 @@
+
+/*******************************************************************************
+* Copyright contributors to the oneDAL project
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+*     http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*******************************************************************************/
+
+#include "oneapi/dal/detail/profiler.hpp"
+
+#include "oneapi/dal/backend/primitives/distance/correlation_distance_misc.hpp"
+
+#include "oneapi/dal/backend/primitives/blas.hpp"
+#include "oneapi/dal/backend/primitives/reduction.hpp"
+#include "oneapi/dal/backend/primitives/stat/cov.hpp"
+
+namespace oneapi::dal::backend::primitives {
+
+template <typename Float, ndorder order>
+sycl::event compute_deviation(sycl::queue& q,
+                              const ndview<Float, 2, order>& inp,
+                              ndview<Float, 2>& out,
+                              const event_vector& deps) {
+    ONEDAL_ASSERT(inp.has_data());
+    ONEDAL_ASSERT(out.has_mutable_data());
+    const auto n = out.get_dimension(0);
+    const auto p = out.get_dimension(1);
+    ONEDAL_ASSERT(n == inp.get_dimension(0));
+    ONEDAL_ASSERT(p == inp.get_dimension(1));
+    auto* const inp_ptr = inp.get_mutable_data();
+    const auto out_stride = out.get_leading_stride();
+    auto* const out_ptr = out.get_mutable_data();
+    auto out_range = make_range_2d(n, p);
+    auto inp_sum = ndarray<Float, 1>::empty(q, { n });
+    auto inp_mean = ndarray<Float, 1>::empty(q, { n });
+
+    // Collect sums of each row of input matrix
+    auto sum_event = reduce_by_rows(q, inp, inp_sum, sum<Float>{}, identity<Float>{}, deps);
+
+    // Compute mean of each row of input matrix using sum event
+    auto mean_event = means(q, p, inp_sum, inp_mean, { sum_event });
+
+    // Return event that updates output matrix with centered values (input(x) - input_mean(x))
+    return q.submit([&](sycl::handler& h) {
+        h.depends_on({ mean_event });
+        auto const inp_mean_acc = inp_mean.get_data();
```

Comment:
I guess more common usage of get_data is before q.submit, and inp_mean_ptr is more common naming for such things